torch/csrc/jit/function_schema.h

```diff
@@ -4,22 +4,28 @@
 
 namespace torch { namespace jit {
 
-struct AttributeInfo {
-  AttributeKind kind;
-  at::optional<int32_t> data; // extra data field, current only used for the k in IntList[k]
-};
-
 // schema as used in the compiler for resolving function calls and reporting
 // errors. These objects should be constructed from C10 schema once those
 // are availiable
 struct Argument {
-  const std::string name;
-  const TypePtr type;
+  Argument(
+      std::string name = "",
+      TypePtr type = nullptr,
+      at::optional<int32_t> N = at::nullopt,
+      at::optional<at::Tensor> default_value = at::nullopt,
+      bool kwarg_only = true)
+      : name(std::move(name)),
+        type(type),
+        N(N),
+        default_value(default_value),
+        kwarg_only(kwarg_only) {}
+  std::string name;
+  TypePtr type;
+  at::optional<int32_t> N; // a fixed length for list types, at::nullopt means the list is dynamically sized
```

Comment:
The way `int[3]` is used in the project makes the `3` more a part of the function declaration than part of the list type for the argument.  In particular, `int[3]` doesn't mean 'I expect a list whose size is statically 3', but instead, 'If someone tries to pass an int here, you can make it into an appropriate list by repeating it 3 times'. If we tried to encode it into the type, I would expect it to have the former meaning and to prevent using a list of unknown length in its place. By putting N in the argument list rather than the type, it is clear it affects the behavior of the call site but still accepts normal lists as arguments.